sources/core/Stride.Core.Mathematics.Tests/TestMatrix.cs

```diff
@@ -0,0 +1,109 @@
+            var rollRadians = MathUtil.DegreesToRadians(rollDegrees);
+
+            var rotMatrix = Matrix.RotationYawPitchRoll(yawRadians, pitchRadians, rollRadians);
+            rotMatrix.Decompose(out float decomposedYaw, out float decomposedPitch, out float decomposedRoll);
+
+            var expectedQuat = Quaternion.RotationYawPitchRoll(yawRadians, pitchRadians, rollRadians);
+            var decompedQuat = Quaternion.RotationYawPitchRoll(decomposedYaw, decomposedPitch, decomposedRoll);
+            Assert.True(expectedQuat == decompedQuat || expectedQuat == -decompedQuat, $"Quat not equals: Expected: {expectedQuat} - Actual: {decompedQuat}");
+        }
+
+        [Theory, MemberData(nameof(RotationAngles))]
+        public void TestDecomposeYawPitchRollFromMatricesZXY(float yawDegrees, float pitchDegrees, float rollDegrees)
+        {
+            var yawRadians = MathUtil.DegreesToRadians(yawDegrees);
+            var pitchRadians = MathUtil.DegreesToRadians(pitchDegrees);
+            var rollRadians = MathUtil.DegreesToRadians(rollDegrees);
+
+            // Yaw-Pitch-Roll is the intrinsic rotation order, so extrinsic is the reverse (ie. Z-X-Y)
+            var rotMatrix = Matrix.RotationZ(rollRadians) * Matrix.RotationX(pitchRadians) * Matrix.RotationY(yawRadians);
+            rotMatrix.Decompose(out float decomposedYaw, out float decomposedPitch, out float decomposedRoll);
+
+            var expectedQuat = Quaternion.RotationYawPitchRoll(yawRadians, pitchRadians, rollRadians);
+            var decompedQuat = Quaternion.RotationYawPitchRoll(decomposedYaw, decomposedPitch, decomposedRoll);
+            Assert.True(expectedQuat == decompedQuat || expectedQuat == -decompedQuat, $"Quat not equals: Expected: {expectedQuat} - Actual: {decompedQuat}");
+        }
+
+        [Theory, MemberData(nameof(RotationAngles))]
+        public void TestDecomposeXYZFromMatricesXYZ(float yawDegrees, float pitchDegrees, float rollDegrees)
+        {
+            var yawRadians = MathUtil.DegreesToRadians(yawDegrees);
+            var pitchRadians = MathUtil.DegreesToRadians(pitchDegrees);
+            var rollRadians = MathUtil.DegreesToRadians(rollDegrees);
+
+            var rotMatrix = Matrix.RotationX(pitchRadians) * Matrix.RotationY(yawRadians) * Matrix.RotationZ(rollRadians);
+            rotMatrix.DecomposeXYZ(out Vector3 eulerAngles);
+
+            var decompedRotMatrix = Matrix.RotationX(eulerAngles.X) * Matrix.RotationY(eulerAngles.Y) * Matrix.RotationZ(eulerAngles.Z);
+            var decompedQuat = Quaternion.RotationMatrix(decompedRotMatrix);
+
+            var expectedQuat = Quaternion.RotationX(pitchRadians) * Quaternion.RotationY(yawRadians) * Quaternion.RotationZ(rollRadians);
+            Assert.True(expectedQuat == decompedQuat || expectedQuat == -decompedQuat, $"Quat not equals: Expected: {expectedQuat} - Actual: {decompedQuat}");
+        }
+
+        public static IEnumerable<object[]> RotationAngles
```

Comment:
While I don't object testing a lot of the cases from a moral standpoint, this producing over 3k test cases may kill Visual Studio test viewer - I'd recommend checking which values hit which code pathways to ensure 100% coverage of logic in those method and keep a few example per pathway.

Reply:
Ok, I've sat down and rethought this. I only really need to concern myself with when the singularity (ie. edge cases occur), which is +/-90 degrees on the second rotation. Other angles are for coverage.